lib/node_modules/@stdlib/iter/while-each/lib/main.js

```diff
@@ -0,0 +1,155 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var setReadOnly = require( '@stdlib/utils/define-nonenumerable-read-only-property' );
+var isFunction = require( '@stdlib/assert/is-function' );
+var isIteratorLike = require( '@stdlib/assert/is-iterator-like' );
+var iteratorSymbol = require( '@stdlib/symbol/iterator' );
+var format = require( '@stdlib/string/format' );
+
+
+// MAIN //
+
+/**
+* Returns an iterator which invokes a function for each iterated value before returning the iterated value.
+*
+* @param {Iterator} iterator - input iterator
+* @param {Function} predicate - function which indicates whether to continue iterating
+* @param {Function} fcn - function to invoke
+* @param {*} [thisArg] - execution context
+* @throws {TypeError} first argument must be an iterator protocol-compliant object
+* @throws {TypeError} second argument must be a function
+* @throws {TypeError} third argument must be a function
+* @returns {Iterator} iterator
+*
+* @example
+* var randu = require( '@stdlib/random/iter/randu' );
+* var isnan = require( '@stdlib/math/base/assert/is-nan' );
+*
+* function predicate( v ) {
+*     return ( v <= 0.5 );
+* }
+*
+* function assert( v ) {
+*     if ( isnan( v ) ) {
+*         throw new Error( 'should not be NaN' );
+*     }
+* }
+*
+* var iter = iterWhileEach( randu(), predicate, assert );
+*
+* var r = iter.next().value;
+* // returns <number>
+*
+* r = iter.next().value;
+* // returns <number>
+*
+* // ...
+*/
+function iterWhileEach( iterator, predicate, fcn, thisArg ) {
+	var iter;
+	var FLG;
+	var i;
+	if ( !isIteratorLike( iterator ) ) {
+		throw new TypeError( format( 'invalid argument. First argument must be an iterator protocol-compliant object. Value: `%s`.', iterator ) );
+	}
+	if ( !isFunction( predicate ) ) {
+		throw new TypeError( format( 'invalid argument. Second argument must be a function. Value: `%s`.', predicate ) );
+	}
+	if ( !isFunction( fcn ) ) {
+		throw new TypeError( format( 'invalid argument. Third argument must be a function. Value: `%s`.', fcn ) );
+	}
+	i = -1;
+
+	// Create an iterator protocol-compliant object:
+	iter = {};
+	setReadOnly( iter, 'next', next );
+	setReadOnly( iter, 'return', end );
+
+	// If an environment supports `Symbol.iterator`, make the iterator iterable:
+	if ( iteratorSymbol && isFunction( iterator[ iteratorSymbol ] ) ) {
+		setReadOnly( iter, iteratorSymbol, factory );
+	}
+	return iter;
+
+	/**
+	* Returns an iterator protocol-compliant object containing the next iterated value.
+	*
+	* @private
+	* @returns {Object} iterator protocol-compliant object
+	*/
+	function next() {
+		var v;
+		i += 1;
+		if ( FLG ) {
+			return {
+				'done': true
+			};
+		}
+		v = iterator.next();
+		if ( v.done ) {
+			FLG = true;
+			return v;
+		}
+		v = v.value;
+		fcn.call( thisArg, v, i );
+		return {
+			'value': v,
+			'done': !predicate( v )
+		};
```

Comment:
```suggestion
		if ( v.done || !predicate( v, i ) {
			FLG = true;
			return v;
		}
		v = v.value;
		fcn.call( thisArg, v, i );
		return {
			'value': v,
			'done': false
		};
```

As currently implemented, `fcn` is being called **before** `predicate` which is not what should happen (see [`@stdlib/utils/while-each`](https://github.com/stdlib-js/stdlib/blob/30cf48578491d545c528a28ccc031545c8e7bbf9/lib/node_modules/%40stdlib/utils/while-each/lib/main.js#L71)). We want to stop calling `fcn` as soon as `predicate` returns `false`; at which point, the iterator is considered "done".

Additionally, as stated in the docs, `predicate` should be provided both the value and index.